src/openfermion/measurements/_bayesian_estimators.py

```diff
@@ -0,0 +1,520 @@
+    """
+
+    def __init__(self,
+                 prior_amplitude_mean=None,
+                 prior_amplitude_var=None,
+                 num_vectors=1,
+                 num_freqs=1000,
+                 max_rotations=1,
+                 prior_vector=None,
+                 full_update_with_failure=False,
+                 store_history=True,
+                 amplitude_approx_cutoff=100):
+
+        """
+        Args:
+            prior_amplitude_mean (1-d numpy array or None):
+                estimates for amplitudes of different
+                eigenstates in the initial state ($A_j$)
+            prior_amplitude_var (2-d numpy array or None):
+                variance on the amplitude estimates
+            num_vectors (int): number of phases to estimate
+            num_freqs (int): number of frequencies in Fourier
+                representation.
+            max_rotations (int): maximum number of unitary
+                evolutions before measurement. Dictates how many
+                matrices for updates are made and stored.
+            prior_vector (2-d numpy array or None):
+                a prior estimate of the phases. If None, assumes flat.
+            full_update_with_failure (bool): chooses whether to perform
+                a full single step of SLSQP whenever the amplitude
+                optimization returns an unphysical result (negative
+                amplitude squared), or to just enforce physicality
+                by hand.
+            store_history (bool): whether to store the history of
+                estimated values of the estimator. This is only
+                a few numbers per update, but it could be costly.
+            amplitude_approx_cutoff (int): a cutoff between 
+                estimating the amplitudes via full SLSQP and
+                approximating with single steps of Newton's
+                method. Increasing may lead to higher accuracy,
+                decreasing will lead to faster runtimes.
+        """
+        self._amplitude_approx_cutoff = amplitude_approx_cutoff
+        self._full_update_with_failure = full_update_with_failure
+        self._store_history = store_history
+        self._probability_dist = FourierProbabilityDist(
+            amplitude_mean=prior_amplitude_mean,
+            amplitude_var=prior_amplitude_var,
+            num_vectors=num_vectors,
+            num_freqs=num_freqs,
+            fourier_vector=prior_vector)
+
+        # Store the initial priors given for use in resetting
+        self._prior_vector = copy.deepcopy(prior_vector)
+        self._prior_amplitude_mean = copy.deepcopy(prior_amplitude_mean)
+        self._prior_amplitude_var = copy.deepcopy(prior_amplitude_var)
+
+        self._make_projector()
+        self._max_rotations = max_rotations
+        self._matrices = make_fourier_matrices(max_rotations, num_freqs)
+
+        self.reset()
+
+    def reset(self):
+        """
+        Resets estimator to initial state.
+        """
```

Comment:
Can you explain why one would ever want to reset the estimator to its initial state?

Reply:
For testing the estimator performance one needs to iterate over multiple estimations and average the result. But, as you say, it would be possible to just create a new estimator for each iteration. When I've been doing this before this was a pain (because creating BayesEstimator._matrices took so long).  But, as I mentioned above, these matrices shouldn't even be in the estimator, as they're unique to FourierProbabilityDist.

As these matrices are static, I think the solution is to store them separately to both classes in an object that gets referenced by every instance of FourierProbabilityDist. This is a bit abstract for me, so I'll have to look around for the most appropriate pythonic way of doing this. Once I've sorted this out, I think I agree with removing the reset function from BayesEstimator.

Reply:
> Thanks for updating the docstring of FourierProbabilityDist. I think I understand now what it is supposed to represent, and it makes sense to have a class for it.
> 
> > I see what you mean here, but I can imagine at some point wanting to replace the FourierProbabilityDist class with another representation of the probability distribution. Which isn't to say we should leave it as is, rather, I think BayesianEstimator needs to be rewritten to treat FourierProbabilityDist as a more abstract object, and FourierProbabilityDist needs to be rewritten to understand its own math.
> > This means (if I'm not mistaken) I need to pull these update matrices out of both BayesianEstimator (as they're distribution-specific) and FourierProbabilityDist.
> > Does that make sense? This will take a fair amount of refactoring, not sure if I can get around to it before the end of the year.
> 
> I would avoid abstracting the probability distribution concept until we actually have another use case. However, it might be good to abstract the estimator concept. Do you ultimately plan to have BayesianEstimator and TimeSeriesEstimator inherit from the same class?

Yes, this would be the ultimate plan. There are a few key differences between the two though, most notably that BayesEstimator can use external data to start with a prior (TimeSeriesEstimator cannot easily), and that TimeSeriesEstimator supports a much more compressed data input (BayesEstimator really needs to process each experiment individually).